gradle-twirl/src/main/java/play/twirl/gradle/TwirlPlugin.java

```diff
@@ -44,24 +47,33 @@ public TwirlPlugin(ObjectFactory objectFactory) {
   public void apply(final Project project) {
     project.getPluginManager().apply(ScalaBasePlugin.class);
 
-    Configuration twirlConfiguration = createDefaultTwirlConfiguration(project);
+    TwirlExtension twirlExtension = project.getExtensions().create("twirl", TwirlExtension.class);
+    twirlExtension.getScalaVersion().convention(DEFAULT_SCALA_VERSION);
+
+    Configuration twirlConfiguration = createDefaultTwirlConfiguration(project, twirlExtension);
 
     configureSourceSetDefaults(project, twirlConfiguration);
   }
 
-  private Configuration createDefaultTwirlConfiguration(Project project) {
+  private Configuration createDefaultTwirlConfiguration(
+      Project project, TwirlExtension twirlExtension) {
     // Get Twirl version from Gradle Plugin MANIFEST.MF
     String twirlVersion = getClass().getPackage().getImplementationVersion();
     Configuration conf = project.getConfigurations().create("twirl");
     conf.setDescription("The Twirl compiler library.");
     conf.setVisible(false);
     conf.setTransitive(true);
-    conf.defaultDependencies(
-        dependencies ->
-            dependencies.add(
-                project
-                    .getDependencies()
-                    .create("com.typesafe.play:twirl-compiler_2.13:" + twirlVersion)));
+    project.afterEvaluate(
+        __ -> {
+          Dependency twirlCompiler =
+              project
+                  .getDependencies()
+                  .create(
+                      String.format(
+                          "com.typesafe.play:twirl-compiler_%s:%s",
+                          twirlExtension.getScalaVersion().get(), twirlVersion));
+          conf.defaultDependencies(dependencies -> dependencies.add(twirlCompiler));
```

Comment:
have you tried moving `dependencies.create(...)` into `defaultDependencies` lambda?
Then `afterEvaluate` will not be required.